alacritty/src/input.rs

```diff
@@ -735,6 +745,118 @@ impl<T: EventListener, A: ActionContext<T>> Processor<T, A> {
+        }
+    }
+
+    /// Handle beginning of touch input.
+    pub fn on_touch_start(&mut self, touch: TouchEvent) {
+        let touch_purpose = self.ctx.touch();
+        *touch_purpose = match mem::take(touch_purpose) {
+            Touch::None => Touch::Tap(touch),
+            Touch::Tap(start) => Touch::Zoom(TouchZoom::new((start, touch))),
+            Touch::Zoom(zoom) => Touch::Invalid(zoom.slots()),
+            Touch::Scroll(event) | Touch::Select(event) => {
+                let mut set = HashSet::new();
+                set.insert(event.id);
+                Touch::Invalid(set)
+            },
+            Touch::Invalid(mut slots) => {
+                slots.insert(touch.id);
+                Touch::Invalid(slots)
+            },
+        };
+    }
+
+    /// Handle touch input movement.
+    pub fn on_touch_motion(&mut self, touch: TouchEvent) {
+        let touch_purpose = self.ctx.touch();
+        match touch_purpose {
+            Touch::None => (),
+            // Handle transition from tap to scroll/select.
+            Touch::Tap(start) => {
+                let delta_x = touch.location.x - start.location.x;
+                let delta_y = touch.location.y - start.location.y;
+                if delta_x.abs() > MAX_TAP_DISTANCE {
+                    // Update gesture state.
+                    let start_location = start.location;
+                    *touch_purpose = Touch::Select(*start);
+
+                    // Start simulated mouse input.
+                    self.mouse_moved(start_location);
+                    self.mouse_input(ElementState::Pressed, MouseButton::Left);
+
+                    // Apply motion since touch start.
+                    self.on_touch_motion(touch);
+                } else if delta_y.abs() > MAX_TAP_DISTANCE {
+                    // Update gesture state.
+                    *touch_purpose = Touch::Scroll(*start);
+
+                    // Apply motion since touch start.
+                    self.on_touch_motion(touch);
+                }
+            },
+            Touch::Zoom(zoom) => {
+                let font_delta = zoom.font_delta(touch);
+                self.ctx.change_font_size(font_delta);
+            },
+            Touch::Scroll(last_touch) => {
+                // Calculate delta and update last touch position.
+                let delta_y = touch.location.y - last_touch.location.y;
+                *touch_purpose = Touch::Scroll(touch);
+
+                self.scroll_terminal(0., delta_y * TOUCH_SCROLL_FACTOR);
+            },
+            Touch::Select(_) => self.mouse_moved(touch.location),
+            Touch::Invalid(_) => (),
+        }
+    }
+
+    /// Handle end of touch input.
+    pub fn on_touch_end(&mut self, touch: TouchEvent) {
+        // Finalize the touch motion up to the release point.
+        self.on_touch_motion(touch);
+
+        let touch_purpose = self.ctx.touch();
+        match touch_purpose {
```

Comment:
```suggestion
        let touch_purpose = self.ctx.touch();
        *touch_purpose = match touch_purpose {
```

And then simply return from each branch. Should do the same for `on_touch_motion`, since you're already doing so in `on_touch_start`.

Reply:
Won't work due to borrow-checker issues. You can't hold the touch purpose across `self.mouse_input` since it's a mutable borrow.

Reply:
But you've already done it before with `std::mem::take`?

Reply:
No, mem::take does not circumvent the borrowchecker issues, it just makes it so you don't have to clone anything. You're still holding `touch_purpose` until reassignment mutably.

Reply:
Then you should change the other instance where you use `mem::take` for that?